dbms/src/Storages/MergeTree/MergeTreeReaderCompact.cpp

```diff
@@ -0,0 +1,272 @@
+        {
+            auto position = data_part->getColumnPosition(part_column.name);
+            if (position && Nested::extractTableName(part_column.name) == table_name)
+                return position;
+        }
+    }
+
+    return {};
+}
+
+
+void MergeTreeReaderCompact::readData(
+    const String & name, IColumn & column, const IDataType & type,
+    size_t from_mark, size_t column_position, size_t rows_to_read, bool only_offsets)
+{
+    if (!isContinuousReading(from_mark, column_position))
+        seekToMark(from_mark, column_position);
+
+    auto buffer_getter = [&](const IDataType::SubstreamPath & substream_path) -> ReadBuffer *
+    {
+        if (only_offsets && (substream_path.size() != 1 || substream_path[0].type != IDataType::Substream::ArraySizes))
+            return nullptr;
+
+        return data_buffer;
+    };
+
+    IDataType::DeserializeBinaryBulkSettings deserialize_settings;
+    deserialize_settings.getter = buffer_getter;
+    deserialize_settings.avg_value_size_hint = avg_value_size_hints[name];
+    deserialize_settings.position_independent_encoding = true;
+
+    IDataType::DeserializeBinaryBulkStatePtr state;
+    type.deserializeBinaryBulkStatePrefix(deserialize_settings, state);
+    type.deserializeBinaryBulkWithMultipleStreams(column, rows_to_read, deserialize_settings, state);
+
+    /// The buffer is left in inconsistent state after reading single offsets
+    if (only_offsets)
+        last_read_granule.reset();
+    else
+        last_read_granule.emplace(from_mark, column_position);
+}
+
+
+void MergeTreeReaderCompact::initMarksLoader()
+{
+    if (marks_loader.initialized())
+        return;
+
+    size_t columns_num = data_part->getColumns().size();
+
+    auto load = [this, columns_num](const String & mrk_path) -> MarkCache::MappedPtr
+    {
+        size_t file_size = Poco::File(mrk_path).getSize();
+        size_t marks_count = data_part->getMarksCount();
+        size_t mark_size_in_bytes = data_part->index_granularity_info.getMarkSizeInBytes(columns_num);
+
+        size_t expected_file_size = mark_size_in_bytes * marks_count;
+        if (expected_file_size != file_size)
+            throw Exception(
+                "Bad size of marks file '" + mrk_path + "': " + std::to_string(file_size) + ", must be: " + std::to_string(expected_file_size),
+                ErrorCodes::CORRUPTED_DATA);
+
+        /// Memory for marks must not be accounted as memory usage for query, because they are stored in shared cache.
+        auto temporarily_disable_memory_tracker = getCurrentMemoryTrackerActionLock();
+
+        auto res = std::make_shared<MarksInCompressedFile>(marks_count * columns_num);
+
+        ReadBufferFromFile buffer(mrk_path, file_size);
+        size_t i = 0;
+
+        while (!buffer.eof())
+        {
+            buffer.readStrict(reinterpret_cast<char *>(res->data() + i * columns_num), sizeof(MarkInCompressedFile) * columns_num);
```

Comment:
Slightly ugly, maybe move this code as independent function to `MarkInCompressedFile`.